atc/engine/engine.go

```diff
@@ -29,12 +31,14 @@ type Runnable interface {
 //go:generate counterfeiter . StepBuilder
 
 type StepBuilder interface {
-	BuildStep(db.Build) (exec.Step, error)
+	BuildStep(db.Build, vars.CredVarsTracker) (exec.Step, error)
```

Comment:
:thinking: I wonder if we could push `creds.Secrets` and `redactSecrets` into the `StepBuilder` instead?

The `StepBuilder` could construct the `vars.CredVarsTracker` on its own since it's already receiving the `db.Build`:

```go
vars.NewCredVarsTracker(creds.NewVariables(secrets, build.TeamName(), build.PipelineName()), redactSecrets)
```

Reply:
The problem is, we want an `vars.CredVarsTracker` object to be global to a build. If each step constructs a `vars.CredVarsTracker` object, then we cannot solve the case like a task interpolates some secrets then pass secrets to second task and second task print secrets.

Reply:
Good point. I revisited the code again, and realized you are totally right. Constructing a tracker in `BuildStep` is more elegant. Have removed tracker from `engine.go` and moved to `builder.go`.